superdesk/core/http/types.py

```diff
@@ -0,0 +1,272 @@
+        methods: Optional[List[HTTP_METHOD]] = None,
+        name: Optional[str] = None,
+    ):
+        self.url = url
+        self.func = func
+        self.methods = methods or ["GET"]
+        self.name = name or func.__name__
+
+    def __call__(self, args: Dict[str, Any], params: Dict[str, Any], request: "HTTPRequest"):
+        func_params = signature(self.func).parameters
+        if "args" not in func_params and "params" not in func_params:
+            return self.func(request)  # type: ignore[call-arg,arg-type]
+
+        arg_type = func_params["args"] if "args" in func_params else None
+        request_args = None
+        if arg_type is not None and arg_type.annotation is not None and issubclass(arg_type.annotation, BaseModel):
+            request_args = arg_type.annotation.model_validate(args)
+
+        param_type = func_params["params"] if "params" in func_params else None
+        url_params = None
+        if (
+            param_type is not None
+            and param_type.annotation is not None
+            and issubclass(param_type.annotation, BaseModel)
+        ):
+            url_params = param_type.annotation.model_validate(params)
+
+        return self.func(request_args, url_params, request)  # type: ignore[call-arg,arg-type]
+
+
+class HTTPRequest(Protocol):
+    """Protocol to define common request functionality
+
+    This is implemented in `SuperdeskEve` app using Flask to provide the required functionality.
+    """
+
+    #: The current HTTPEndpoint being processed
+    endpoint: HTTPEndpoint
+
+    @property
+    def method(self) -> HTTP_METHOD:
+        """Returns the current HTTP method for the request"""
+        ...
+
+    @property
+    def path(self) -> str:
+        """Returns the URL of the current request"""
+        ...
+
+    def get_header(self, key: str) -> Optional[str]:
+        """Get an HTTP header from the current request"""
+        ...
+
+    async def get_json(self) -> Union[Any, None]:
+        """Get the body of the current request in JSON format"""
+        ...
+
+    async def get_form(self) -> Mapping:
+        """Get the body of the current request in form format"""
+        ...
+
+    async def get_data(self) -> Union[bytes, str]:
+        """Get the body of the current request in raw bytes format"""
+        ...
+
+    async def abort(self, code: int, *args: Any, **kwargs: Any) -> NoReturn:
+        ...
+
+
+class HTTPEndpointGroup:
+    """Base class used for registering a group of endpoints"""
+
+    #: Optional url prefix to be added to all routes of this group
+    url_prefix: Optional[str]
+
+    #: List of endpoints registered with this group
+    endpoints: List[HTTPEndpoint]
+
+    def __init__(self, url_prefix: Optional[str] = None):
+        self.url_prefix = url_prefix
+        self.endpoints = []
+
+    def endpoint(
+        self,
+        url: str,
+        name: Optional[str] = None,
+        methods: Optional[List[HTTP_METHOD]] = None,
```

Comment:
would be good to handle authentication/authorization here as well

Reply:
Yes that's on my list. Auth (including RBAC) and projection is not currently implemented. I'll create 2 tickets for those